### ISSUE TITLE
Enums are defined, but not used internally. Replace with string constants?

### ISSUE THREAD

**Issue by MichaelClerx** [cellml/libcellml#203]
`Variable.setInterfaceType()` must be called with an InterfaceType, but the get() method returns a string. I'm sure it's convenient for the xml file, but surely that shouldn't be implemented here?

It should be possible to do

    v.setInterfaceType(v.getInterfaceType())

or

    v1.setInterfaceType(v2.getInterfaceType())

for example when cloning a variable

**Comment by agarny** [comment]
There are two methods for setting an interface type:
 - `void Variable::setInterfaceType(const std::string &interfaceType)`; and
 - `void Variable::setInterfaceType(Variable::InterfaceType interfaceType)`.

So, I am not sure what your issue is here...?

**Comment by MichaelClerx** [comment]
Woops, I didn't see the `set(string)` method!

Regardless though. Why do we define an interface type enum and then not use it? Why do we provide a method that lets you pass in any old garbage? Why are we storing a string internally? The whole point of having an API is that it's something more than the XML file

**Comment by agarny** [comment]
Yes, I agree that from an enduser's perspective, it would be nice to have something like `Variable::InterfaceType getInterfaceType() const`.

About having the interface type set to any string value, well... I know, but I will leave it to @nickerso to justify that decision... :)

**Comment by hsorby** [comment]
This sort of API (the enumeration I am thinking of here) is provided for application developers so that they can fill entries in a list box user interface widget and similar devices.

With C++ you can differentiate methods on parameter but you cannot differentiate on return type, thus you have to choose the one.  There is obviously going to be some trade offs with this.  We have to remember that we are not against users putting together invalid models, for instance in this case setting the interface type to `bob`.  What we supply is a `Validator` that checks the validity of the model and advises about errors detected.

If you look at the tests you will see the expected behaviour, indeed you should look at the tests to see if how you intend (are actually) using the library is the way it actually behaves.  The library code is a result of the tests so that is where you should look when thinking is this what I want from this library.

When I look at the tests I see the behaviour that I want from the library and so think the API is correct, certainly for this case.

**Comment by agarny** [comment]
Good advice, @hsorby, about checking the tests. I should do it more often myself!

**Comment by MichaelClerx** [comment]
We don't have to guarantee it's valid, but we should make it easy to write valid models and difficult to write invalid ones.
With this API, the user has to _remember_ what the valid strings are that will be returned. Why don't we use the enum? That's what enums are for! At the very least we could replace it with defined string constants. This is madness.

**Comment by nickerso** [comment]
In order to cope with the range of expected input and keep the API and data model as simple as possible, the variable type property must be a string. Without introducing exceptions, I don't see a way that we can have a `Variable::InterfaceType Variable::getInterfaceType()` - apart from introducing the singularly unhelpful `Variable::InterfaceType::UNKNOWN` :)

If people see a need for the currently internal `interfaceTypeToString` map to be visible via the API, then I don't currently see a problem with that.

**Comment by MichaelClerx** [comment]
I just saw this in `Units.cpp`:

    void Units::addUnit(const std::string &reference, Prefix prefix, double exponent, double multiplier)
    {
        auto search = prefixToString.find(prefix);
        assert(search != prefixToString.end());
        const std::string prefixString = search->second;
        addUnit(reference, prefixString, exponent, multiplier);
    }

So internally, we're converting enums to strings. Would it be a good idea to remove enums from the API completely in favour of string constants? This would save us a lookup here, and would also allow a call like `addUnit('newton', 2)` to be mapped to `addUnit(string, double)` instead of `addUnit(string, Prefix)` which is probably not what the user intended. Looking at the other enums I've encountered so far I can see no reason not to replace them all by string constants...

It would also solve this issue, as it would make it clear to the user that strings are used internally, and the enum is just a convenience thing that will get converted to a string by the API

**Comment by MichaelClerx** [comment]
It would also make these things a little simpler:

    addUnit('newton', 1) maps to addUnit('newton', Prefix)
    addUnit('newton', 1.0) maps to addUnit('newton', double exponent)

If a prefix was always a string, it would be obvious that the second call didn't specify a prefix...

**Comment by MichaelClerx** [comment]
Update: Just realised the issue above might only occur in the Python bindings. Since Python doesn't have enums it stores them as integer constants, and the method accepting an enum becomes a method accepting any int...

Still think it would be good to have string constants instead of enums though (ideally enums that _are_ string constants, but C++ doesn't appear to allow that)

**Comment by MichaelClerx** [comment]
Enum issues also mean we can't have both a `Units.removeUnit(size_t)` and a `Units.removeUnit(StandardUnit)` in the Python bindings.

(What's the use case for `removeUnit(StandardUnit)` and `removeUnit(reference)` anyway? Feels a bit random to me)
